### ISSUE TITLE
consequences of tokenizer divergence between TransformersLens and Huggingface Transformers

### ISSUE THREAD

**Issue by randommm** [TransformerLensOrg/TransformerLens#1418]
I don't know if this is intended and I couldn't find much about it, it seems that the tokenizers of TransformersLens add a BOS token by default when encoding while Transformers doesn't.

I know this can be disabled with `skip_special_tokens`, but my concern is whether this is the case when we can call `model.generate` with a prompt as text, adding an unexpected `bos` breaking the model expected first token distribution silently.

From what I understand here, it will call `to_tokens` which will call `tokenizer.encode` without `add_special_tokens=False`.

So if my understanding is correct, this is seems like a footgun. Could someone clarify this behavior?

Maybe a nice to have is some special argument on generate that have it return the tokens that it used to generate to help debugging and making sure nothing unexpected was added.

**Code example**
```python
import torch
from transformer_lens.model_bridge.bridge import TransformerBridge
from transformers import AutoModelForCausalLM, AutoTokenizer

prompt = "2 + 2?"

MODEL = "allenai/OLMoE-1B-7B-0125-Instruct"
model = TransformerBridge.boot_transformers(
    MODEL,
    dtype=torch.bfloat16,
    device_map="auto",
    trust_remote_code=False,
)
model.eval()
tl_tokenizer = model.tokenizer

hf_tokenizer = AutoTokenizer.from_pretrained(MODEL)

v1 = tl_tokenizer.apply_chat_template(
    [{"role": "user", "content": prompt}],
    add_generation_prompt=True,
    return_tensors="pt",
    tokenize=True,
)["input_ids"][0].tolist()
v2 = tl_tokenizer.encode(
    tl_tokenizer.apply_chat_template(
        [{"role": "user", "content": prompt}],
        add_generation_prompt=True,
        return_tensors="pt",
        tokenize=False,
    )
)
print("Token check for transformer lens is", v1 == v2)
print(v1[:3], v2[:4])

v1 = hf_tokenizer.apply_chat_template(
    [{"role": "user", "content": prompt}],
    add_generation_prompt=True,
    return_tensors="pt",
    tokenize=True,
)["input_ids"][0].tolist()
v2 = hf_tokenizer.encode(
    hf_tokenizer.apply_chat_template(
        [{"role": "user", "content": prompt}],
        add_generation_prompt=True,
        return_tensors="pt",
        tokenize=False,
    )
)
print("Token check for hf transformers is", v1 == v2)
```

Output:
```
Token check for transformer lens is False
[50279, 29, 93] [50279, 50279, 29, 93]
Token check for hf transformers is True
```

**Comment by jlarson4** [comment]
Hi @randommm! Thanks for the write up on this. There are two behaviors tangled together, one is intended and one is the real footgun.

## The breakdown

TransformerLens makes two deliberate choices about BOS:

1. **`model.tokenizer` is loaded with `add_bos_token=True`** (`sources/transformers.py`), so the tokenizer auto-prepends BOS. This is an intentional divergence stock HF tokenizer. We modify it so that direct tokenizer use matches TL's BOS-by-default convention.
2. **`cfg.default_prepend_bos = True`**, because many interpretability results rely on a token at position 0.

Inside `to_tokens` (which is what `forward`/`generate` use for string input), these are reconciled to exactly one BOS: if the tokenizer doesn't auto-prepend, one is added; if you pass `prepend_bos=False`, the auto-added one is stripped.

So there are two cases:

1.  **Going around `to_tokens`** (calling `model.tokenizer.encode(...)` directly, as in your repro): you get TL's raw `add_bos_token=True` behavior, which will diverge from stock HF and can double-add on text that already contains a BOS. This is by design, but it is definitely under-documented.

2.  **Within `to_tokens`, on chat models**: this is the real footgun you're pointing at. When you pass text from `apply_chat_template(..., tokenize=False)`, the rendered string already embeds a BOS as text. `to_tokens`' reconciliation only normalizes the tokenizer's auto-prepend versus the flag. It doesn't detect a BOS already sitting in the input text, so you end up with two BOS tokens, silently shifting the first-token distribution. That's a legitimate problem.

## The good news: the existing knob already fixes the chat case

You can already get the correct, HF-matching result today:

```python
model.to_tokens(templated_text, prepend_bos=False)   # or generate(..., prepend_bos=False)
```

With `prepend_bos=False`, `to_tokens` strips exactly one leading BOS, leaving the single BOS that the chat template embedded. The gap is awareness and visibility, rather than capability.

## The Resolution

Whoever would like to tackle this issue (whether its you or another contributor), lets implement the following to ensure it is easier to find these answers.

- [ ] **Add a `return_tokens` flag to surface the tokens that were actually used** (your "nice to have"). Whether `prepend_bos=False` is the right choice depends on whether a given model's chat template embeds a BOS, which is model-specific. Add a way to return/inspect the input tokens from the generation path (alongside the fact that `model.to_tokens(prompt, prepend_bos=...)` already returns exactly what `generate` will feed). 
- [ ] **Document the BOS contract explicitly**, in the `to_tokens` / "Tokenization notes" section add the following comment:
```
# model.tokenizer is configured with add_bos_token=True and is **not** the stock HF tokenizer. Direct .encode() calls will prepend BOS.
# When passing pre-applied chat-template text (apply_chat_template(..., tokenize=False)), pass prepend_bos=False to avoid a double BOS.
```


TL;DR – BOS-by-default is intentional (and controllable), the raw-tokenizer divergence is by design, but undocumented, and the chat double-BOS is a real footgun that `prepend_bos=False` exists to fixes. We'll add token visibility + docs so it's no longer silent.

**Comment by MdSadiqMd** [comment]
Hey @jlarson4 can i work on this issue please

**Comment by jlarson4** [comment]
@MdSadiqMd certainly!

**Comment by jlarson4** [comment]
@randommm A little update on this: The `return_tokens` flag has been added, along with a comment explaining the `prepend_bos` configuration. `generate` and `generate_stream` have also been updated to ensure that they respect the `prepend_bos` value when it is provided. Let me know if you have any other questions, I am going to mark this as closed.
